### PR TITLE
cli: Rework commands

### PR DESCRIPTION
**Pull request by ernestask**
This commit makes all commands loadable at runtime using an abstraction
that should make adding command-line arguments easier, meaning that this
also replaces any and all uses of python-argh to work directly with
argparse. This is needed to have the flexibility required to create
fringe subparsers on demand and better handle things like exclusive
argument groups.

Functionality-wise, this doubles down on the “filtering” concept: any
commands that operates on given problems will further filter the
problems based on other arguments, e.g. --since, --component, -N, etc.

--since, --until, -N, --component and --executable are all available for
use to:

    - filter based on time of occurrence,
    - select N latest problems,
    - filter based on problem component,
    - filter based on executable.

Fixes https://github.com/abrt/abrt/issues/291
Fixes https://github.com/abrt/abrt/issues/293

Signed-off-by: Ernestas Kulik <ekulik@redhat.com>

### CODE REVIEW

**Comment by ernestask** [comment]
I need to see if the integration tests need adjusting (although changes should be backwards-compatible for the most part). Otherwise, this is good for reviewing.

File: src/cli/abrtcli/cli/__init__.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    def parser(self):
+        return self._parser
+
+    def add_filter_arguments(self):
+        group = self._parser.add_argument_group()
+
+        def uint(string):
+            index = int(string)
+            if index < 1:
+                raise argparse.ArgumentTypeError(_('positive non-zero integer expected'))
```
Comment (mkutlak) [inline src/cli/abrtcli/cli/__init__.py]: Undefined variable `argparse`.

File: src/cli/abrtcli/cli/install_debuginfo.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            problem.chown()
+
+            with remember_cwd():
+                try:
+                    os.chdir(problem.path)
+                except OSError:
+                    print(_('Permission denied: \'{}\'\n'
+                            'If this is a system problem'
+                            ' try running this command as root')
+                          .format(prob.path))
```
Comment (mkutlak) [inline src/cli/abrtcli/cli/install_debuginfo.py]: Undefined variable `prob`.

File: src/cli/abrtcli/cli/install_debuginfo.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                                      executables=arguments.executables,
+                                      components=arguments.components,
+                                      since=arguments.since,
+                                      until=arguments.until,
+                                      n_latest=arguments.n_latest,
+                                      not_reported=arguments.not_reported)
+        for problem in problems:
+            if not isinstance(problem, Ccpp):
+                print(_('The problem is not of a C/C++ type. Can\'t install debuginfo'))
+                sys.exit(1)
```
Comment (mkutlak) [inline src/cli/abrtcli/cli/install_debuginfo.py]: Undefined variable `sys`.

File: src/cli/abrtcli/cli/report.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                                      authenticate=arguments.authenticate,
+                                      executables=arguments.executables,
+                                      components=arguments.components,
+                                      since=arguments.since,
+                                      until=arguments.until,
+                                      n_latest=arguments.n_latest,
+                                      not_reported=arguments.not_reported)
+        for problem in problems:
+            if problem.not_reportable and not arguments.unsafe:
+                if reportclient.verbose > 0:
```
Comment (mkutlak) [inline src/cli/abrtcli/cli/report.py]: Undefined variable `reportclient`.

File: src/cli/abrtcli/cli/report.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                    print(problem.not_reportable_reason)
+
+                print(_('Problem \'{0}\' cannot be reported').format(problem.short_id))
+                sys.exit(1)
```
Comment (mkutlak) [inline src/cli/abrtcli/cli/report.py]: Undefined variable `sys`.

File: src/cli/abrtcli/cli/status.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+from abrtcli import cli, config
+from abrtcli.match import match_lookup
+from abrtcli.utils import format_problems
```
Comment (mkutlak) [inline src/cli/abrtcli/cli/status.py]: Unused import `format_problems`.

File: src/cli/abrtcli/cli/__init__.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+from abrtcli import config
+from abrtcli.i18n import _
+from reportclient import set_verbosity
+
+class Command:
+    aliases = []
+
+    def __init__(self, subparsers):
+        self._parser = subparsers.add_parser(self.name, aliases=self.aliases,
+                help=self.description, description=self.description)
```
Comment (mkutlak) [inline src/cli/abrtcli/cli/__init__.py]: No-member `name`, `description`.

File: src/cli/abrtcli/cli/install_debuginfo.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                                      authenticate=arguments.authenticate,
+                                      executables=arguments.executables,
+                                      components=arguments.components,
+                                      since=arguments.since,
+                                      until=arguments.until,
+                                      n_latest=arguments.n_latest,
+                                      not_reported=arguments.not_reported)
+        for problem in problems:
+            if not isinstance(problem, Ccpp):
+                print(_('The problem is not of a C/C++ type. Can\'t install debuginfo'))
```
Comment (mkutlak) [inline src/cli/abrtcli/cli/install_debuginfo.py]: Undefined variable `_`.

File: src/cli/abrtcli/cli/list.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
+from abrtcli import cli, config
+from abrtcli.i18n import _
+from abrtcli.match import match_lookup
+from abrtcli.utils import format_problems, sort_problems
```
Comment (mkutlak) [inline src/cli/abrtcli/cli/list.py]: Unused import `sort_problems`.

File: src/cli/abrtcli/cli/retrace.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+                                      since=arguments.since,
+                                      until=arguments.until,
+                                      n_latest=arguments.n_latest,
+                                      not_reported=arguments.not_reported)
+        for problem in problems:
+            if hasattr(problem, 'backtrace') and not force:
+                print(_('Problem already has a backtrace'))
+                print(_('Run abrt retrace with -f/--force to retrace again'))
+                if ask_yes_no(_('Show backtrace?')):
+                    print(format_problems(problem, format=config.BACKTRACE_FMT))
```
[the remark this answers is not in the document]
Comment (mkutlak) [reply src/cli/abrtcli/cli/retrace.py]: Undefined variable `config`.

File: src/cli/abrt [hunk 11]
```diff
@@ -1,17 +1,46 @@
 #!/usr/bin/python3
 
+import argparse
 import sys
 import subprocess
```
Comment (mkutlak) [inline src/cli/abrt]: Unused import `subprocess`.

File: src/cli/abrtcli/cli/__init__.py (hunk 7 above)
Comment (ernestask) [reply src/cli/abrtcli/cli/__init__.py]: Hmm, this was more or less intended, but probably not a good idea.

File: src/cli/abrtcli/cli/list.py (hunk 9 above)
Comment (ernestask) [reply src/cli/abrtcli/cli/list.py]: Not sure why that is, because it’s absolutely used.

Comment (ernestask) [reply src/cli/abrtcli/cli/list.py]: Oh, `sort_problems`. Me blind.

**Comment by ernestask** [comment]
That was quite embarrassing.

**Comment by mkutlak** [comment]
`pylint` is complaining about some of the indentations and order of imports (everything from `problem`, `reportclient` should be placed before `abrtcli` imports).

File: configure.ac [hunk 12]
```diff
@@ -437,6 +437,7 @@ AC_CONFIG_FILES([
 	src/applet/Makefile
 	src/cli/Makefile
 	src/cli/abrtcli/Makefile
+    src/cli/abrtcli/cli/Makefile
```
Comment (mgrabovsky) [inline configure.ac]: A tab instead of spaces would be consistent here.

**Comment by ernestask** [comment]
Myeah, I need to go back to the drawing board after seeing the output from pylint.

**Comment by mkutlak** [comment]
@trapas test this please

**Comment by ernestask** [comment]
@traps retest this please

File: src/cli/tests/test_utils.py [hunk 13]
```diff
@@ -54,48 +54,48 @@ def test_fmt_problems(self):
...
 
-    def test_fmt_problems_oneline(self):
+    def test_format_problems_oneline(self):
         '''
         Test oneline problem formatting
         '''
 
         pl = problem.list()
-        res = fmt_problems(pl, fmt=ONELINE_FMT)
+        res = format_problems(pl, format=ONELINE_FMT)
```
Comment (mkutlak) [inline src/cli/tests/test_utils.py]: TypeError: format_problems() got an unexpected keyword argument 'format'

**Comment by ernestask** [comment]
@trapas test this please

**Comment by ernestask** [comment]
@trapas retest this please

**Comment by ernestask** [comment]
@trapas test this please

**Comment by ernestask** [comment]
@trapas test this please

**Comment by ernestask** [comment]
Test failures were due to build files being completely broken. Not sure why it stopped working, but, essentially, the tests were being run against the system installation, which isn’t what you want to do in unit-testing.

**Comment by mkutlak** [comment]
```
positional arguments:
PATTERN               path to the problem directory, problem ID or ->wildcard<-
```
```

$ abrt info 
Id            4c319ad

$ abrt info *
No matching problems found

$ abrt info "*"
Id            4c319ad

$ abrt info "4*"
No matching problems found
```
Is it really a "wildcard" if only thing that works is `"*"`?

**Comment by ernestask** [comment]
> ```
> positional arguments:
> PATTERN               path to the problem directory, problem ID or ->wildcard<-
> ```
> 
> ```
> 
> $ abrt info 
> Id            4c319ad
> 
> $ abrt info *
> No matching problems found
> 
> $ abrt info "*"
> Id            4c319ad
> 
> $ abrt info "4*"
> No matching problems found
> ```
> 
> Is it really a "wildcard" if only thing that works is `"*"`?

1 is your shell doing globbing, so it will naturally not work. Go to /var/spool/abrt and try that again, the results will be __shocking__.
2 is the intended usage, unless you configure your shell to do something else to allow 1.
3 is misinterpreting the description. I thought of allowing regexes or simple globbing, but the concept of filtering made more sense to me, as otherwise it’s a bit tricky to get it to play nicely with other options (not to mention the problem in 1). If you have some ideas as to what would work best, throw them at me.

**Comment by ernestask** [comment]
fwiw, bash autocompletion should help a bit, but I didn’t really test it, since it does not work with zsh out of the box.

**Comment by mkutlak** [comment]
All commands now run for ALL problems.

`abrt rm` - tries to remove all problems
`abrt gdb` - opens gdb for all problems
`abrt info` - shows info for all problems
`abrt di` - install debuginfo for all problems
`abrt retrace` - runs retracing for all problems

Bug or feature?

(Before only the latest problem was used.)

**Comment by ernestask** [comment]
Very much a bug, it should be using the latest one.

**Comment by ernestask** [comment]
I thought I had that implemented, but seems that I’m already creating false memories. Should be good now.

**Comment by mkutlak** [comment]
LGTM.

**Pull request merged**